### DISCUSSION TITLE
Unable to run visualization properly!

### DISCUSSION THREAD

**Discussion by ghost** [mesa/mesa#1749] [Q&A] [answered]
I had written my code a month ago and everything worked fine. I installed the latest version of Mesa 2.1.1 today and I have been facing issues with the visualization. I am not sure if installing the upgrade is what caused the issue. I tried to run the boid flockers example and it worked fine. 

<img width="1595" alt="Screenshot 2023-08-02 at 9 57 15 AM" src="https://github.com/projectmesa/mesa/assets/131771458/5ecb2756-8c88-4515-b049-5577b9af1349">

The error log I see in "spyder" is as follows:

WARNING:tornado.general:403 GET /local/SimpleCanvas//Users/preetamkulkarni/Library/CloudStorage/OneDrive-UniversityofTexasatArlington/NSF%20CAREER%20-%20Project/Research%20Paper%20Submissions/WSC%202023/Code/Python%20MESA%20code/simple_continuous_canvas.js (127.0.0.1): /Users/preetamkulkarni/Library/CloudStorage/OneDrive-UniversityofTexasatArlington/NSF CAREER - Project/Research Paper Submissions/WSC 2023/Code/Python MESA code/simple_continuous_canvas.js is not in root static directory
WARNING:tornado.access:403 GET /local/SimpleCanvas//Users/preetamkulkarni/Library/CloudStorage/OneDrive-UniversityofTexasatArlington/NSF%20CAREER%20-%20Project/Research%20Paper%20Submissions/WSC%202023/Code/Python%20MESA%20code/simple_continuous_canvas.js (127.0.0.1) 15.32ms
Socket opened!
{"type":"reset"}
100

WARNING:tornado.general:403 GET /local/SimpleCanvas//Users/preetamkulkarni/Library/CloudStorage/OneDrive-UniversityofTexasatArlington/NSF%20CAREER%20-%20Project/Research%20Paper%20Submissions/WSC%202023/Code/Python%20MESA%20code/simple_continuous_canvas.js (127.0.0.1): /Users/preetamkulkarni/Library/CloudStorage/OneDrive-UniversityofTexasatArlington/NSF CAREER - Project/Research Paper Submissions/WSC 2023/Code/Python MESA code/simple_continuous_canvas.js is not in root static directory
WARNING:tornado.access:403 GET /local/SimpleCanvas//Users/preetamkulkarni/Library/CloudStorage/OneDrive-UniversityofTexasatArlington/NSF%20CAREER%20-%20Project/Research%20Paper%20Submissions/WSC%202023/Code/Python%20MESA%20code/simple_continuous_canvas.js (127.0.0.1) 15.32ms

**Comment by ghost** [comment]
I was able to resolve this by adding the proper address of the file "**_simple_continuous_canvas.js_**" in "**_SimpleContinuousModule.py_**". This was my mistake because I had moved files around and I forgot to update.

However, I do have this issue with the legends and colors not showing up on the screen correctly after I updated to Mesa 2.1.1. 

<img width="1595" alt="Screenshot 2023-08-02 at 12 12 15 PM" src="https://github.com/projectmesa/mesa/assets/131771458/7dfa256d-d790-41e9-837a-b9600bae8be6">

Please help me resolve this. 

Thanks in advance :)

**Reply by rht** [reply]
If I read this correctly, the original error is not due to the 2.1.1 release. But with the new error, you haven't specified any console error log.

**Reply by ghost** [reply]
Yes, You are right about the original error. It had nothing to do with mesa 2.1.1. The second error is new and I have not changed anything in the code from the time it was working fine. There is nothing in the error log that points to the issue as far as I am understanding. Please correct me if I am wrong. Below is the error log (I can also send you my code on your email ID. Please let me know.)

runfile('/Users/preetamkulkarni/Library/CloudStorage/OneDrive-UniversityofTexasatArlington/NSF CAREER - Project/Research Paper Submissions/JASSS - DACE and ABM/Python MESA code/server.py', wdir='/Users/preetamkulkarni/Library/CloudStorage/OneDrive-UniversityofTexasatArlington/NSF CAREER - Project/Research Paper Submissions/JASSS - DACE and ABM/Python MESA code')
100
Interface starting at http://127.0.0.1:8521
Traceback (most recent call last):

  File ~/Library/CloudStorage/OneDrive-UniversityofTexasatArlington/NSF CAREER - Project/Research Paper Submissions/JASSS - DACE and ABM/Python MESA code/server.py:76 in <module>
    server.launch()

  File ~/opt/anaconda3/lib/python3.9/site-packages/mesa/visualization/ModularVisualization.py:407 in launch
    tornado.ioloop.IOLoop.current().start()

  File ~/opt/anaconda3/lib/python3.9/site-packages/tornado/platform/asyncio.py:195 in start
    self.asyncio_loop.run_forever()

  File ~/opt/anaconda3/lib/python3.9/asyncio/base_events.py:591 in run_forever
    self._check_running()

  File ~/opt/anaconda3/lib/python3.9/asyncio/base_events.py:583 in _check_running
    raise RuntimeError('This event loop is already running')

RuntimeError: This event loop is already running

Socket opened!
{"type":"reset"}
100
{"type":"get_step","step":1}
{"type":"get_step","step":2}
{"type":"get_step","step":3}
{"type":"get_step","step":4}
{"type":"get_step","step":5}
{"type":"get_step","step":6}
{"type":"get_step","step":7}
{"type":"get_step","step":8}
{"type":"get_step","step":9}
{"type":"get_step","step":10}
{"type":"get_step","step":11}
{"type":"get_step","step":12}
{"type":"get_step","step":13}
{"type":"get_step","step":14}
{"type":"get_step","step":15}
{"type":"get_step","step":16}
{"type":"get_step","step":17}
{"type":"get_step","step":18}
{"type":"get_step","step":19}
{"type":"get_step","step":20}
{"type":"get_step","step":21}
{"type":"get_step","step":22}
{"type":"get_step","step":23}
{"type":"get_step","step":24}

Here is a screenshot of when I run my model for reference. As seen in the picture, the legend and the color are missing.

<img width="1595" alt="Screenshot 2023-08-03 at 9 59 30 AM" src="https://github.com/projectmesa/mesa/assets/131771458/6322fdd2-4f20-4f73-8b1e-3077233a7816">

I also ran the "Geo Schelling" model which I downloaded from the github repository and it also has the same issue as shown below:

<img width="1595" alt="Screenshot 2023-08-03 at 10 03 33 AM" src="https://github.com/projectmesa/mesa/assets/131771458/16e26050-2e2b-4ad2-aebf-353bb2f83810">

**Reply by rht** [reply]
- Have you tried downgrading to 1.x, and confirmed the problem disappears?
- The error message indicates there is already an existing Tornado server running. Have you made sure your session is fresh and clean before retrying?

**Reply by ghost** [reply]
Hi,
I used "ctrl + ." before running code and this restarts the kernal. I also tried closing spyder IDE completely and restarted it. I also removed mesa 2.1.1 and installed 1.2.1 but it is still not showing the legend and the color on the graphs. I am not sure what is causing the issue. I checked the version and it shows I have downgraded to a lower version:

pip show mesa
Name: Mesa
Version: 1.2.1
Summary: Agent-based modeling (ABM) in Python 3+
Home-page: https://github.com/projectmesa/mesa
Author: Project Mesa Team
Author-email: projectmesa@googlegroups.com
License: Apache 2.0
Location: /Users/preetamkulkarni/opt/anaconda3/lib/python3.9/site-packages
Requires: pandas, tornado, tqdm, cookiecutter, networkx, click, numpy
Required-by: Mesa-Geo
Note: you may need to restart the kernel to use updated packages.

**Reply by rht** [reply]
Can you try running the server from a terminal instead of via Spyder? The Tornado viz framework is known to not work well with Jupyter, because Jupyter has Tornado. I suspect that Spyder has some sort of Jupyter stuff under the hood.

There is also the new Solara viz framework, but there are no visualizers yet for continuous space and geo stuff.

**Reply by ghost** [reply, marked as answer]
I tried to update Anaconda and Spyder and it fixed the issue. I am able to recreate the issue by upgrading to Mesa 2.1.1. 
Everything works fine if I am on Mesa 1.2.1 and for now, I will continue to use Mesa 1.2.1.

<img width="1450" alt="Screenshot 2023-08-03 at 12 20 47 PM" src="https://github.com/projectmesa/mesa/assets/131771458/e109431f-dd8b-4c6c-b1b0-eff765b8f978">

**Reply by NickGotts** [reply]
I have been struggling with a very similar problem. I have been trying to port a continuous space model written under Mesa 0.8.8.1 to Mesa 2.1.1. The model runs (after small changes such as to account for an updated version of shapely), but the visualization does not work. Moreover, after an unsuccessful run, the boid_flockers example sometimes (but not always) behaves in the same way, and I have to close and reopen the browser (Firefox) to get it to do so. Other examples, using grid spaces, are unaffected even when boid_flockers is. I'm running the model from a powershell terminal of an anaconda environment using Python 3.8,, which is the same way I've been running the model throughout its development. Any suggestions, please?

**Reply by rht** [reply]
1. Do you have a minimal reproducer?
2. Can you post error messages in the console that you have encountered?

**Reply by rht** [reply]
> Moreover, after an unsuccessful run, the boid_flockers example sometimes (but not always) behaves in the same way

If I run boid_flockers I should see the problem?

**Reply by NickGotts** [reply]
Thanks for getting back to me so promptly!

My "minimal reproducer" is an early version of my model, which is still not very minimal; I will try cutting it down further. The skeleton of it is based on boid_flockers itself, particularly with regard to the interaction with JavaScript (a language of which I know very little). No error messages appear on the console. However, if I start a Jupyter notebook from within anaconda, navigate to the folder with "run.py" in it and input "import run", I get the following:

Interface starting at http://127.0.0.1:8521/

---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In[1], line 1
----> 1 import run

File ~\anaconda3\envs\mesaenv38new\mesa-examples\examples\distancing2csimp\run.py:3
      1 from distancing2c.server import server  # noqa
----> 3 server.launch(open_browser=True)

File ~\anaconda3\lib\site-packages\mesa\visualization\ModularVisualization.py:407, in ModularServer.launch(self, port, open_browser)
    405 tornado.autoreload.start()
    406 try:
--> 407     tornado.ioloop.IOLoop.current().start()
    408 except KeyboardInterrupt:
    409     tornado.ioloop.IOLoop.current().stop()

File ~\anaconda3\lib\site-packages\tornado\platform\asyncio.py:195, in BaseAsyncIOLoop.start(self)
    194 def start(self) -> None:
--> 195     self.asyncio_loop.run_forever()

File ~\anaconda3\lib\asyncio\base_events.py:560, in BaseEventLoop.run_forever(self)
    558 """Run until stop() is called."""
    559 self._check_closed()
--> 560 self._check_running()
    561 self._set_coroutine_origin_tracking(self._debug)
    562 self._thread_id = threading.get_ident()

File ~\anaconda3\lib\asyncio\base_events.py:552, in BaseEventLoop._check_running(self)
    550 def _check_running(self):
    551     if self.is_running():
--> 552         raise RuntimeError('This event loop is already running')
    553     if events._get_running_loop() is not None:
    554         raise RuntimeError(
    555             'Cannot run the event loop while another loop is running')

RuntimeError: This event loop is already running

Socket opened!
{"type":"reset"}

****

The model runs when started, so the problem seems to be closely tied to the visualization.

As for your query about boid_flockers, I don't know, but I would guess not, since boid_flockers usually works. I am guessing there is something about my model which causes the subsequent failure of the boid_flockers visualization, but what that could be, I don't know, and it does seem to be specific to continuous space, as other example models run correctly even after visualization fails for both my model and boid_flockers. I will see if I can reproduce the problem by running boid_flockers repeatedly, interrupting it, trying to start up a new run while the old one is paused, and so on.

**Reply by rht** [reply]
That's the same error that @PreetamKulkarni encountered. It means there are multiple Tornado servers running. The most common reason is that you are running inside a Jupyter notebook, which itself uses Tornado.

Do you want to try using the new pure Python viz frontend instead? No JS needed. You just have to draw with Matplotlib. It's still new, so I haven't implemented for continuous space yet, but you can see an example of how to draw a custom space in https://github.com/projectmesa/mesa-examples/blob/main/examples/sugarscape_g1mt/app.py. The bonus benefit is that you can run the viz inside Jupyter. Important note when implementing the space drawer is that you must do `ax.scatter` (Matplotlib's OOP API) instead `plt.scatter`.

**Reply by NickGotts** [reply]
That's not the explanation in this case - at least, not the whole explanation, because I've now tried running the same model in the same way under Mesa 0.8.8.1, and the visualization works fine. That environment does have an older version of Jupyter, but since I'm getting the same problem with the visualization under Mesa 2.1.1 whether I use Jupyter or not, and in both cases the problem is absent under version 0.8.8.1, some interaction of my code with the new Mesa version looks like the most likely cause of the problem. I would be interested in trying out a pure Python frontend at some point, but I have a paper to write, so I think I'll have to stick with Mesa 0.8.8.1 and the JavaScript front end for now!
